File: Python/92. Reverse Linked List II.py
# Given the head of a singly linked list and two integers left and right where left <= right, reverse the nodes of the list from position left to position right, and return the reversed list.

 

# Example 1:


# Input: head = [1,2,3,4,5], left = 2, right = 4
# Output: [1,4,3,2,5]
# Example 2:

# Input: head = [5], left = 1, right = 1
# Output: [5]
 

# Constraints:

# The number of nodes in the list is n.
# 1 <= n <= 500
# -500 <= Node.val <= 500
# 1 <= left <= right <= n
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution(object):
    def reverseBetween(self, head, left, right):
        values = []   # to stores the values that needed to be reverse
        nodes = []    # to keep track of the nodes which needs to be reverse
        counter = 0
        curr = head
        while curr:
            counter += 1
            if counter >= left and counter <= right:
                values.append(curr.val)     #  | adding value and nodes in the list
                nodes.append(curr)          #  |
            curr = curr.next
        for node in nodes:
            node.val = values.pop(-1)       #  replace the value of nodes
        return head

File: C++/92. Reverse Linked List II.cpp
// Given the head of a singly linked list and two integers left and right where left <= right, reverse the nodes of the list from position left to position right, and return the reversed list.

 

// Example 1:


// Input: head = [1,2,3,4,5], left = 2, right = 4
// Output: [1,4,3,2,5]
// Example 2:

// Input: head = [5], left = 1, right = 1
// Output: [5]


class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    reverseBetween(head, left, right) {
        let left2right = [];
        let left2rightVal = [];
        let counter = 0;
        let curr = head;
        
        while (curr) {
            counter++;
            if (left <= counter && counter <= right) {
                left2right.push(curr);
                left2rightVal.push(curr.val);
            }
            curr = curr.next;
        }
        
        for (let node of left2right) {
            node.val = left2rightVal.pop();
        }
        
        return head;
    }
}